pyscf/pbc/df/fft_occk.py

```diff
@@ -0,0 +1,269 @@
+    # occ
+    if gamma_point(kpts_band) and gamma_point(kpts):
+        kiv = np.zeros((nset,nband,mo_coeff0[0].shape[1],nao), dtype=dms.dtype)
+    else:
+        kiv = [[]*nset]
+        for i in range(nset):
+            for k1 in range(nband):
+                kiv[i].append(np.zeros((mo_coeff0[k1].shape[1],nao), dtype=np.complex128))
+
+    mem_now = lib.current_memory()[0]
+    max_memory = mydf.max_memory - mem_now
+    blksize = int(min(nao, max(1, (max_memory-mem_now)*1e6/16/4/ngrids/nao)))
+    lib.logger.debug1(mydf, 'fft_jk: get_k_kpts max_memory %s  blksize %d',
+                      max_memory, blksize)
+    ao1_dtype = np.result_type(*ao1_kpts)
+    ao2_dtype = np.result_type(*ao2_kpts)
+    vR_dm = np.empty((nset,nao,ngrids), dtype=vk_kpts.dtype)
+
+    t1 = (time.clock(), time.time())
+    for k2, ao2T in enumerate(ao2_kpts):
+        if ao2T.size == 0:
+            continue
+
+        kpt2 = kpts[k2]
+        naoj = ao2T.shape[0]
+
+        if mo_coeff is None or nset > 1:
+            ao_dms = [lib.dot(dms[i,k2], ao2T.conj()) for i in range(nset)]
+        else:
+            ao_dms = [ao2T.conj()]
+        for k1, ao1T in enumerate(ao1_kpts):
+            kpt1 = kpts_band[k1]
+            # occ
+            ao3T = np.dot(mo_coeff0[k1].T, ao1T)
+            # If we have an ewald exxdiv, we add the G=0 correction near the
+            # end of the function to bypass any discretization errors
+            # that arise from the FFT.
+            mydf.exxdiv = exxdiv
+            if exxdiv == 'ewald' or exxdiv is None:
+                coulG = tools.get_coulG(cell, kpt2-kpt1, False, mydf, mesh)
+            else:
+                coulG = tools.get_coulG(cell, kpt2-kpt1, True, mydf, mesh)
+            if is_zero(kpt1-kpt2):
+                expmikr = np.array(1.)
+            else:
+                expmikr = np.exp(-1j * np.dot(coords, kpt2-kpt1))
+
+            m3T = ao3T.shape[0]
+            for p0, p1 in lib.prange(0, m3T, blksize):
+                rho1 = np.einsum('ig,jg->ijg', ao3T[p0:p1].conj()*expmikr, ao2T)
+                vG = tools.fft(rho1.reshape(-1,ngrids), mesh)
+                rho1 = None
+                vG *= coulG
+                vR = tools.ifft(vG, mesh).reshape(p1-p0,naoj,ngrids)
+                vG = None
+                if vR_dm.dtype == np.double:
+                    vR = vR.real
+                for i in range(nset):
+                  np.einsum('ijg,jg->ig', vR, ao_dms[i], out=vR_dm[i,p0:p1])
+                vR = None
+            vR_dm *= expmikr.conj()
+            for i in range(nset):
+                # occ
+                kiv[i][k1] += weight * lib.dot(vR_dm[i,0:mo_coeff[k1].shape[1]], ao1T.T)
+
+        t1 = lib.logger.timer_debug1(mydf, 'get_k_kpts: make_kpt (%d,*)'%k2, *t1)
+    for i in range(nset):
+        for k1, ao1T in enumerate(ao1_kpts):
+            kij = lib.einsum('ui,ju->ij', mo_coeff0[k1], kiv[i][k1])
+            kr = scipy.linalg.solve(kij.conj(), kiv[i][k1])
```

Comment:
Let's try Cholesky decomposition for `-kij` first. If it fails, do an eigen decomposition and eliminate the eigenvectors corresponding to the small eigenvalues.